From 00dc5a807a9b3e608ac2134b6cbe2e589790fd52 Mon Sep 17 00:00:00 2001
From: Jiri Olsa <jolsa@redhat.com>
Date: Tue, 19 Aug 2014 15:25:29 -0400
Subject: [tools] perf/trace: Allow specifying index offset in strarrays

Message-id: <1408462094-14194-161-git-send-email-jolsa@redhat.com>
Patchwork-id: 88117
O-Subject: [PATCH RHEL7.1 BZ1131394 160/325] perf trace: Allow specifying index offset in strarrays
Bugzilla: 1131394
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>

From: Jiri Olsa <jolsa@kernel.org>

Bugzilla: 1131394
https://bugzilla.redhat.com/show_bug.cgi?id=1131394

upstream
========
commit 03e3adc9f4d8b57dc83475c8c4c6e462a78ff709
Author: Arnaldo Carvalho de Melo <acme@redhat.com>
Date: Tue Oct 8 16:00:21 2013 -0300

description
===========
So that the index passed doesn't have to start at zero, being
decremented from an offset specified when declaring the strarray before
being used as the real array index.
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/tools/perf/builtin-trace.c b/tools/perf/builtin-trace.c
index 0c46af7..e1fa7cc 100644
--- a/tools/perf/builtin-trace.c
+++ b/tools/perf/builtin-trace.c
@@ -44,6 +44,7 @@ struct syscall_arg {
 };
 
 struct strarray {
+ int     offset;
  int     nr_entries;
  const char **entries;
 };
@@ -53,14 +54,20 @@ struct strarray {
  .entries = array, \
 }
 
+#define DEFINE_STRARRAY_OFFSET(array, off) struct strarray strarray__##array = { \
+ .offset     = off, \
+ .nr_entries = ARRAY_SIZE(array), \
+ .entries = array, \
+}
+
 static size_t syscall_arg__scnprintf_strarray(char *bf, size_t size,
            struct syscall_arg *arg)
 {
- int idx = arg->val;
  struct strarray *sa = arg->parm;
+ int idx = arg->val - sa->offset;
 
  if (idx < 0 || idx >= sa->nr_entries)
-  return scnprintf(bf, size, "%d", idx);
+  return scnprintf(bf, size, "%d", arg->val);
 
  return scnprintf(bf, size, "%s", sa->entries[idx]);
 }
@@ -288,8 +295,8 @@ static size_t syscall_arg__scnprintf_futex_op(char *bf, size_t size, struct sysc
 
 #define SCA_FUTEX_OP  syscall_arg__scnprintf_futex_op
 
-static const char *epoll_ctl_ops[] = { [1] = "ADD", "DEL", "MOD", };
-static DEFINE_STRARRAY(epoll_ctl_ops);
+static const char *epoll_ctl_ops[] = { "ADD", "DEL", "MOD", };
+static DEFINE_STRARRAY_OFFSET(epoll_ctl_ops, 1);
 
 static const char *itimers[] = { "REAL", "VIRTUAL", "PROF", };
 static DEFINE_STRARRAY(itimers);
-- 
1.7.1